src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -216,25 +216,36 @@ public CopyPosition CopyTo(CopyPosition position, T[] array, int arrayIndex, int
             int row = position.Row;
             int column = position.Column;
 
-            for (; count > 0; row++, column = 0)
+            T[] buffer = GetBuffer(row);
+            int copied = CopyToCore(buffer, column);
+
+            while (count > 0)
+            {
+                buffer = GetBuffer(++row);
+                copied = CopyToCore(buffer, 0);
+            }
+
+            return copied == buffer.Length
+                ? new CopyPosition(row + 1, 0)
+                : new CopyPosition(row, copied);
```

Comment:
Hmm, I believe the previous setup I had in place which looked like

```cs
/* first copy */
if (count == 0)
{
    return new CopyPosition(row, column).Normalize(buffer.Length);
}

do
{
    /* subsequent copies */
}
while (count > 0);

return new CopyPosition(row, column).Normalize(buffer.Length);
```

was correct. With the current code, something like

```cs
new TE<int>(new[] { 1 }).Concat(new TE<int>(new[] { 2 })).Concat(new TE<int>(new[] { 3 })).Concat(new[] { 4 }).ToArray()
```

or

```cs
new[] { 1, 2, 3, 4 }.SelectMany(i => i == 4 ? new[] { i } : new TE<int>(new[] { i })).ToArray()
```

should result in `1 2 2 4`. Can you verify that's true?

If it is, you can add a regression test for them named `NonCollectionThatDoesNotBeginAtStartOfRow_FollowedByOtherNonCollections`. What happens is the LAB the non-collections are buffered into looks like this:

```
    C0 C1 C2
R0:  1  2  3
```

The first LAB.CopyTo takes (row = 0, column = 0) as input, copies 1 item (`1`), and returns (0, 1). The 2nd LAB.CopyTo takes (0, 1) as input, copies 1 item (`2`) and returns (0, 1) when it should return (0, 2). The 3rd LAB.CopyTo then takes (0, 1) as input and copies `2` instead of `3`.

Reply:
The previous code is not needed because we now assert that `count` is never zero. However, I'll add the tests mentioned.